src/Client/HedgedConnections.cpp

```diff
@@ -0,0 +1,509 @@
+            next_replica_in_process = false;
+        }
+
+        hedged_connections_factory.stopChoosingReplicas();
+    }
+}
+
+std::string HedgedConnections::dumpAddresses() const
+{
+    std::lock_guard lock(cancel_mutex);
+
+    std::string addresses;
+    bool is_first = true;
+
+    for (const auto & offset_state : offset_states)
+    {
+        for (const auto & replica : offset_state.replicas)
+        {
+            if (replica.connection)
+            {
+                addresses += (is_first ? "" : "; ") + replica.connection->getDescription();
+                is_first = false;
+            }
+        }
+    }
+
+    return addresses;
+}
+
+void HedgedConnections::sendCancel()
+{
+    std::lock_guard lock(cancel_mutex);
+
+    if (!sent_query || cancelled)
+        throw Exception("Cannot cancel. Either no query sent or already cancelled.", ErrorCodes::LOGICAL_ERROR);
+
+    for (auto & offset_status : offset_states)
+        for (auto & replica : offset_status.replicas)
+            if (replica.connection)
+                replica.connection->sendCancel();
+
+    cancelled = true;
+}
+
+Packet HedgedConnections::drain()
+{
+    std::lock_guard lock(cancel_mutex);
+
+    if (!cancelled)
+        throw Exception("Cannot drain connections: cancel first.", ErrorCodes::LOGICAL_ERROR);
+
+    Packet res;
+    res.type = Protocol::Server::EndOfStream;
+
+    while (!epoll.empty())
+    {
+        Packet packet = receivePacketImpl();
+        switch (packet.type)
+        {
+            case Protocol::Server::PartUUIDs:
+            case Protocol::Server::Data:
+            case Protocol::Server::Progress:
+            case Protocol::Server::ProfileInfo:
+            case Protocol::Server::Totals:
+            case Protocol::Server::Extremes:
+            case Protocol::Server::EndOfStream:
+                break;
+
+            case Protocol::Server::Exception:
+            default:
+                /// If we receive an exception or an unknown packet, we save it.
+                res = std::move(packet);
+                break;
+        }
+    }
+
+    return res;
+}
+
+Packet HedgedConnections::receivePacket()
+{
+    std::lock_guard lock(cancel_mutex);
+    return receivePacketUnlocked({});
+}
+
+Packet HedgedConnections::receivePacketUnlocked(AsyncCallback async_callback)
+{
+    if (!sent_query)
+        throw Exception("Cannot receive packets: no query sent.", ErrorCodes::LOGICAL_ERROR);
+    if (!hasActiveConnections())
+        throw Exception("No more packets are available.", ErrorCodes::LOGICAL_ERROR);
+
+    if (epoll.empty())
+        throw Exception("No pending events in epoll.", ErrorCodes::LOGICAL_ERROR);
+
+    return receivePacketImpl(std::move(async_callback));
+}
+
+Packet HedgedConnections::receivePacketImpl(AsyncCallback async_callback)
+{
+    int event_fd;
+    Packet packet;
+    bool finish = false;
+    while (!finish)
+    {
+        event_fd = getReadyFileDescriptor(async_callback);
```

Comment:
Here we have an event loop.
It may be a little bit surprisingly if you read this code the first time.
What we actually need here is a replica `location`, from were we can read packet (call `receivePacketFromReplica`).
So, maybe it is better to extract separate method, like `processEpollDescriptors` or `messageLoop` or something. Which just returns `location` or throws.